### ISSUE TITLE
created imageurl from blob which is from indexeddb is not working in offline

### ISSUE THREAD

**Issue by kamalakkanni** [ujjwalguptaofficial/JsStore#20]
Hi,

I have created blobs from imageurl and got imageurl from blob then stored those in indexeddb using jsstore.Then I fetched the imageurl which is fetched from blob and assigned src of img.its working fine in online.It is not working in offline .img is blinking the app

#https://stackoverflow.com/questions/48580980/fetched-imageurl-is-not-working-in-offline

Thanks

**Comment by ujjwalguptaofficial** [comment]
Ok, I will look into this. Probably I will share some code snippet.

Thanks.

**Comment by kamalakkanni** [comment]
Hi,

This is just reminder for you.Kindly provide me soln as soon as possible

**Comment by kamalakkanni** [comment]
hi,storing image as blob or string which is better?

**Comment by ujjwalguptaofficial** [comment]
Hi - storing image is never a good idea. It will take huge size and in the end it will make your app slow.

Now consider the situation - where people are opening your app in mobile phones. Lets say a user has mobile phone with 1 gb ram and 4 gb storage (there are still some users) and they have very less amount of storage available. They open your app and you filled the storage with blob, although browser will throw the error but yes you need to take care of these situation. User with low end devices will have bad reputation of your app.

So it will be better to not store too many of blob. If you are looking for showing images too in offline, you can use service worker and cache the images and then you can show that in offline.

Thanks

**Comment by ujjwalguptaofficial** [comment]
Some case for storing images will be  - like storing user's profile pic, some company logo and others small pic. I think you are storing too many of dynamic pic which is not a good idea.

**Comment by kamalakkanni** [comment]
I used sw precache to generate service worker automatically.Can I cache
dynamic images using sw precache and service worker

**Comment by ujjwalguptaofficial** [comment]
I have never done this, but yes that should work.

The concept will be - to when you will pass the image url through service worker, you need to fetch the response for the image from the cache.

You gotta do little research and some hit and trial.

**Comment by ujjwalguptaofficial** [comment]
btw - the optimization tip for storing image will be to not define the column in jsstore and store the data.

When you don't define the column - no indexes will be issue for the column and you can't search data for that column but you can store the column. I think you are not going to search on blob column, so it will make your app little faster.

**Comment by kamalakkanni** [comment]
Thanks,But now also it is slow.Fetching blob in offline is still blinking
the app.

**Comment by kamalakkanni** [comment]
Manager is asking for blob only.I can fetch image from by assigning blob
url to image src using blob which is from indexeddb.It is fine in
online.But I don't know why its blinking the app in offline and not
fetching the image

**Comment by kamalakkanni** [comment]
If this issue is fixed means ,i will show it in offline and then I will
tell him about size issue and fetching issue

**Comment by sam-blackfly** [comment]
Hi,
Take a look at this plunkr, this is exactly what you are looking for

[https://plnkr.co/edit/FQJiFMei7ULlDWPntIGy?p=preview]

**Comment by kamalakkanni** [comment]
Thanks,I will try this and let you know

**Comment by kamalakkanni** [comment]
Hi,I have gone through your code.blob is created from file.But I am having
only file url..I have implemented below logic

CreateBlobForImage(imageurl, articleId) {
let self = this;
// Create XHR
var xhr = new XMLHttpRequest(), blob;

xhr.open("GET", imageurl + "_TOPVPF.gif", true);
// Set the responseType to blob
xhr.responseType = "arraybuffer";
xhr.addEventListener("load", function () {
if (xhr.status === 200) {
var arrayBufferView = new Uint8Array(xhr.response);
blob = new Blob([arrayBufferView], { type: "image/gif" });
// Get window.URL object
var url = window.URL;
var imgURL = url.createObjectURL(blob);

// Put the received blob into IndexedDB
self.UpdateBLobInIndexedDb(blob, imgURL, articleId);
}
}, false);
// Send XHR
xhr.send();

}

UpdateBLobInIndexedDb(blob, imgURL, articleId) {
this._connection.openDb(this._indexedDBLiveDBName);
this._connection.update({
In: this._indexedDBLiveDBArticleTableName,
Set: { ImageBlob: blob, ImageBlobUrl: imgURL },
Where: { ArticleId: articleId }
});
}
It is working fine which means blob is created and bloburl fetches images
in online.But Fetching image using bloburl is not working and blinking the
app

**Comment by kamalakkanni** [comment]
But Fetching image using bloburl is not working and blinking the app in
offline only

**Comment by ujjwalguptaofficial** [comment]
you are storing bloburl and using blob url to show the image but you should not store the blob url. You need to create the blob url out of blob. 

The code for getting blob url from blob is - 

```
function getImageUrlFromBlob(file) {
    return new Promise(function (resolve, reject) {
        var reader = new FileReader();
        reader.onload = function (e) {
            resolve(e.target.result);
        };
        reader.onerror = function (error) {
            console.log('Error: ' + error);
            reject(error);
        };
        reader.readAsDataURL(file);
    });

}
```

let me know if this is making sense and its working for you.

**Comment by ujjwalguptaofficial** [comment]
```
var url = window.URL;
var imgURL = url.createObjectURL(blob);

```
don't use this code. Create image url from the blob using the code that I gave you and also don't store it,. You have blob, you can generate the url any time.

Thanks.

**Comment by kamalakkanni** [comment]
ok,thanks

**Comment by kamalakkanni** [comment]
How can I call this and assign to image src while fetching data from
indexeddb.I tried below code

this.GetArticleFromIndexedDb(section, false).then(res => {
let _list=[];
res.forEach(function (item) {
console.log(item.ImageBlob);
this.GetBlobUrlFromBlob(item.ImageBlob).then(res=>{
console.log(res);
_list.push({
Id:item.Id,
Name: item.Name,
.
.
.

});
});
this._article = _list;
});

But I am getting "Promise rejection: Cannot read property
'GetBlobUrlFromBlob' of undefined" error

**Comment by ujjwalguptaofficial** [comment]
use arrow function in forEach like - 
```
res.forEach((item)=>) {
})

```

**Comment by kamalakkanni** [comment]
Thanks its fetches image.But gives Failed to execute 'readAsDataURL' on 'FileReader': parameter 1 is not of type 'Blob'. ; Zone: angular ; Task: Promise.then ; Value: TypeError: Failed to execute 'readAsDataURL' on 'FileReader': parameter 1 is not of type 'Blob'. error sometimes.

**Comment by ujjwalguptaofficial** [comment]
error is clear, that the file you are providing is not blob. Now you need to figure out why and where the things are breaking. 

Sorry, but I cant help you like this, i have very limited time. You can ask such  question on stackoverflow. If there is anything related to jsstore, i would love to solve that but can't solve debugging issue for your particular projects.

Thanks

**Comment by kamalakkanni** [comment]
Thanks.But it fetches image while running ng serve -o.But gives error in production mode (ng build --prod)
 Property 'result' does not exist on type 'EventTarget'.pls help me atleast to solve this

**Comment by kamalakkanni** [comment]
now its not working in ng serve -o also.gives compile error **Property 'result' does not exist on type 'EventTarget'.**

**Comment by ujjwalguptaofficial** [comment]
I don't see property result in the code you have provided.

**Comment by kamalakkanni** [comment]
It is in Creating bloburl
 reader.onload = function (e) {
                    resolve(e.target.result);
                };
which u provide

**Comment by ujjwalguptaofficial** [comment]
Are you still getting the error - 

> parameter 1 is not of type 'Blob'

**Comment by kamalakkanni** [comment]
yes.But, it is running sometimes without giving compile error below

Property 'result' does not exist on type 'EventTarget'.If this compile error not coming means it gives parameter 1 is not of type 'Blob'

**Comment by ujjwalguptaofficial** [comment]
do one thing - at the place of ```e.target.result``` write ```(e as any).target.result``` . This should solve the compilation problem.

**Comment by kamalakkanni** [comment]
ok,thanks
